Only make a shallow copy when loading optimizer state_dict

janeyx99 opened the pull request:
The thing we do still deep copy is the param_groups, which is much lighter weight. This should also save memory when loading from a checkpoint.

The deepcopy was introduced in https://github.com/pytorch/pytorch/commit/ecfcf39f302f7bb193884f72ef2bd59141e5c46c, but module.py had only a shallow copy at that point so it did not actually bring parity.

Incorporates an XLA fix, which is why I'm updating the pin to https://github.com/pytorch/xla/commit/ca5eab87a71f80cd3168630511d02549cc7d2516

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106082

janeyx99 committed:
Only make a shallow copy when loading optimizer state_dict

[ghstack-poisoned]

janeyx99 committed:
Update on "Only make a shallow copy when loading optimizer state_dict"

This should also save memory when loading from a checkpoint.




[ghstack-poisoned]

janeyx99 committed:
Update on "Only make a shallow copy when loading optimizer state_dict"

This should also save memory when loading from a checkpoint.




[ghstack-poisoned]

janeyx99 commented on .git-blame-ignore-revs:
```diff
@@ -30,3 +30,5 @@ e3900d2ba5c9f91a24a9ce34520794c8366d5c54
 # 2023-06-06 clang-format on Foreach / Multi-Tensor-Apply
 515c4279416f13fcc3c898e560f8ae8f15139a03
+# [optim][BE] split test file into logical parts: SWA, LR, optim
+a53cda1ddc15336dc1ff0ce1eff2a49cdc5f882e
```
This could go in another PR but it didn't feel worth another PR....I can be easily convinced otherwise

albanD replied:
TIL this file exists and github uses it by default. Very cool!

albanD reviewed (approved):
SGTM! Very good catch!

janeyx99 committed:
Update on "Only make a shallow copy when loading optimizer state_dict"

The thing we do still deep copy is the param_groups, which is much lighter weight. This should also save memory when loading from a checkpoint.

The deepcopy was introduced in https://github.com/pytorch/pytorch/commit/ecfcf39f302f7bb193884f72ef2bd59141e5c46c, but module.py had only a shallow copy at that point so it did not actually bring parity.




[ghstack-poisoned]

albanD reviewed (approved):
Ok!

janeyx99 committed:
Update on "Only make a shallow copy when loading optimizer state_dict"

The thing we do still deep copy is the param_groups, which is much lighter weight. This should also save memory when loading from a checkpoint.

The deepcopy was introduced in https://github.com/pytorch/pytorch/commit/ecfcf39f302f7bb193884f72ef2bd59141e5c46c, but module.py had only a shallow copy at that point so it did not actually bring parity.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Only make a shallow copy when loading optimizer state_dict"

The thing we do still deep copy is the param_groups, which is much lighter weight. This should also save memory when loading from a checkpoint.

The deepcopy was introduced in https://github.com/pytorch/pytorch/commit/ecfcf39f302f7bb193884f72ef2bd59141e5c46c, but module.py had only a shallow copy at that point so it did not actually bring parity.

Incorporates an XLA fix, which is why I'm updating the pin to https://github.com/pytorch/xla/commit/ca5eab87a71f80cd3168630511d02549cc7d2516




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
